korangar_util/src/container/simple_cache.rs

```diff
@@ -0,0 +1,264 @@
+            max_count,
+            size: AtomicUsize::new(0),
+            max_size,
+        });
+
+        Self {
+            statistics,
+            lookup,
+            values,
+            cache,
+        }
+    }
+
+    /// Returns the statistics of the cache.
+    #[inline(always)]
+    pub fn statistics(&self) -> Arc<Statistics> {
+        self.statistics.clone()
+    }
+
+    /// Returns the maximal count of values inside the cache.
+    #[inline(always)]
+    pub fn max_count(&self) -> u32 {
+        self.statistics.max_count.get()
+    }
+
+    /// Returns the current size of all values inside cache.
+    #[inline(always)]
+    pub fn size(&self) -> usize {
+        self.statistics.size.load(Ordering::Acquire)
+    }
+
+    /// Returns the maximal count of values inside the cache.
+    #[inline(always)]
+    pub fn max_size(&self) -> usize {
+        self.statistics.max_size.get()
+    }
+
+    /// Inserts a value of the given size. The cache saves the given value and
+    /// will drop it when there is not enough size for new cache entries.
+    ///
+    /// If the cache value is too big to be saved inside the cache, this
+    /// function will return a [`ValueTooBig`] error.
+    pub fn insert(&mut self, key: K, value: V) -> Result<(), ValueTooBig> {
+        let size = value.size();
+
+        if size > self.max_size() {
+            return Err(ValueTooBig);
+        }
+
+        let _ = self.remove(&key);
+
+        // Drop as many values as we need to fit the new value.
+        while self.cache.count() > self.statistics.max_count.get().saturating_sub(1)
+            || self.cache.size() > self.statistics.max_size.get().saturating_sub(size)
+        {
+            let (_, key) = self.cache.pop().ok_or(ValueTooBig)?;
+            let cache_key = self.lookup.remove(&key).unwrap();
+            let _ = self.values.remove(cache_key);
+        }
+
+        let cache_key = self.values.insert(value).expect("slab is full");
+        self.lookup.insert(key.clone(), cache_key);
+        self.cache.put(cache_key, key, size);
+        self.update_statistics();
+
+        Ok(())
+    }
+
+    /// Returns a reference of the given cached value.
+    #[must_use]
+    pub fn get<Q>(&mut self, key: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        if let Some(&cache_key) = self.lookup.get(key) {
+            self.cache.touch(cache_key);
+            self.values.get(cache_key)
+        } else {
+            None
+        }
+    }
+
+    /// Removes the value with the given key from the cache.
+    #[must_use]
+    pub fn remove<Q>(&mut self, key: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq + ?Sized,
+    {
+        if let Some(cache_key) = self.lookup.remove(key) {
+            self.cache.remove(cache_key);
+            let value = self.values.remove(cache_key);
+            self.update_statistics();
+            value
+        } else {
+            None
+        }
```

Comment:
I think this could be done more elegantly with `and_then`

Reply:
Fixed